### PR TITLE
Implement osu! hidden mod support

### PR DESCRIPTION
**Pull request by Aergwyn**
I had to redo everything after I broke my branch but here is a "prototype" for HD mod in standard.

I'm not sure if this can change everything that needs to be changed to get it equal to stable but this should be a start that one can work with.

Especially unsure I'm of spinners because they changed a lot compared to stable.

EDIT: [Example Video](https://youtu.be/uBL3sFWF6I8)

- [x] adjust FadeIn and FadeOut to correct values
- [x] change SliderBody to fade out correctly
- [x] look into ApproachCircles not liking to be hurried
- [x] fix Spinners disappearing too early

### CODE REVIEW

**Comment by smoogipoo** [comment]
My initial concern is that you've removed the .Expire()s on every hitobject type, which will cause them to be iterated over even while they're invisible.

**Review by smoogipoo — DISMISSED** [review]

File: osu.Game.Rulesets.Osu/Mods/OsuMod.cs [hunk 1]
```diff
@@ -26,10 +27,43 @@ public class OsuModEasy : ModEasy
 
     }
 
-    public class OsuModHidden : ModHidden
+    public class OsuModHidden : ModHidden, IApplicableToDrawableHitObjects
```
Comment (smoogipoo) [inline osu.Game.Rulesets.Osu/Mods/OsuMod.cs]: Am not really a fan of directly referencing the drawable representations of hit objects and changing their properties. I think better is for the drawable representations to handle their own fadeouts.

How about changing this to inherit from IApplicableToHitObject, and setting a field like `PreEmptFadeOut` on the hitobject?

File: osu.Game.Rulesets.Osu/Mods/OsuMod.cs [hunk 2]
```diff
@@ -26,10 +27,43 @@ public class OsuModEasy : ModEasy
...
+
+                    if (osuDrawable is DrawableSlider slider)
+                    {
+                        slider.GraduallyFadeOut = true;
+
+                        slider.InitialCircle.ShowApproachCircle = false;
+                        slider.InitialCircle.PlayHitAnimation = false;
+                        slider.InitialCircle.FadeInSpeed = osuDrawable.FadeInSpeed;
+                        slider.InitialCircle.FadeOutSpeed = osuDrawable.FadeOutSpeed;
+                        slider.InitialCircle.EarlyFadeOutTime = osuDrawable.EarlyFadeOutTime;
```
Comment (smoogipoo) [inline osu.Game.Rulesets.Osu/Mods/OsuMod.cs]: These should all be handled by the drawable hit objects.

File: osu.Game.Rulesets.Osu/Objects/Drawables/DrawableOsuHitObject.cs [hunk 3]
```diff
@@ -32,20 +38,30 @@ protected sealed override void UpdateState(ArmedState state)
...
         }
 
         protected virtual void UpdateCurrentState(ArmedState state)
         {
         }
 
+        protected virtual void UpdatePostState()
+        {
+            double duration = ((HitObject as IHasEndTime)?.EndTime ?? HitObject.StartTime) - HitObject.StartTime + EarlyFadeOutTime;
+            this.Delay(duration).Expire();
```
Comment (smoogipoo) [inline osu.Game.Rulesets.Osu/Objects/Drawables/DrawableOsuHitObject.cs]: Isn't this going to break for non-hidden for hitcircles? Essentially `StartTime - StartTime + 0`, I think I would prefer that expire was explicitly done down each relevant code path in any case, in-case the transformations change at some point.

Comment (Aergwyn) [reply osu.Game.Rulesets.Osu/Objects/Drawables/DrawableOsuHitObject.cs]: It works fine as I tested so far. It took me a while to understand how and why but it seems to be the `BeginDelayedSequence` that causes it to expire when all Transformations (and Delays) are completed. I didn't notice any cutoffs in normal play.

**Comment by Aergwyn** [comment]
The `.Expire()`s are all handled in `UpdatePostState()` now.

File: osu.Game.Rulesets.Osu/Mods/OsuMod.cs (hunk 1 above)
Comment (Aergwyn) [reply osu.Game.Rulesets.Osu/Mods/OsuMod.cs]: So the idea is to move all new fields I exposed to the hitobject instead? I'll try once at home.

Comment (Aergwyn) [reply osu.Game.Rulesets.Osu/Mods/OsuMod.cs]: Mh. If I move all of them they'd have to be in `OsuHitObject` as I'd have to cast for some fields otherwise. That doesn't sound desirable.
But if I move them all to `OsuHitObject` a field named `ShowApproachCircle` wouldn't make much sense (or one of the other 3).
If I merge them together under a generic name it isn't really specific what that field would do.

So I'd like to keep it as it is. What do you think?

Comment (Aergwyn) [reply osu.Game.Rulesets.Osu/Mods/OsuMod.cs]: @smoogipoo I summon thee for an opinion. So I can continue over the weekend if needed.

Comment (smoogipoo) [reply osu.Game.Rulesets.Osu/Mods/OsuMod.cs]: I don't understand the reason for why it doesn't make sense to put it in the hitobject?

Comment (Aergwyn) [reply osu.Game.Rulesets.Osu/Mods/OsuMod.cs]: This was mostly about those 4 flags that disable or hide something. Some changes are quite specific to the respective type of HitObject. Take hiding the ApproachCircles as an example.
I wouldn't be able to move that one into HitCircle as I'd have to cast the HitObject each time to see if it actually is a HitCircle. Having it in HitObject would work but Spinners and Sliders (except for their InitialCircle) don't have ApproachCircles.
It wouldn't feel right to have them there. Unless I'm overthinking this...

**Comment by peppy** [comment]
Talking with smoogi today, we came to the agreement that this stuff should be done directly on Drawable representations rather than adding extra items to the HitObject model classes. I think this means you may need to revert some of your last commits.

**Comment by peppy** [comment]
Rather than FadeOutGradually on sliders, why not use the same `FadeOutSpeed` and `FadeOutPreempt`? The values can be based off the slider's duration.

File: osu.Game.Rulesets.Osu/Objects/Drawables/DrawableOsuHitObject.cs [hunk 4]
```diff
@@ -14,6 +16,28 @@ public class DrawableOsuHitObject : DrawableHitObject<OsuHitObject>
...
+            set
+            {
+                if (preemptFadeOut == value)
+                    return;
+
+                preemptFadeOut = MathHelper.Clamp(value, 0, 1000);
+            }
+        }
+
+        public override bool IsPresent => base.IsPresent || State.Value == ArmedState.Idle && Time.Current >= HitObject.StartTime - TIME_PREEMPT;
```
Comment (smoogipoo) [inline osu.Game.Rulesets.Osu/Objects/Drawables/DrawableOsuHitObject.cs]: This is pretty weird, why is this needed?

Comment (peppy) [reply osu.Game.Rulesets.Osu/Objects/Drawables/DrawableOsuHitObject.cs]: to stop invisible circles from not receiving input

File: osu.Game.Rulesets.Osu/Objects/Drawables/DrawableHitCircle.cs [hunk 5]
```diff
@@ -88,50 +87,54 @@ protected override void UpdatePreemptState()
...
         {
-            double duration = ((HitObject as IHasEndTime)?.EndTime ?? HitObject.StartTime) - HitObject.StartTime;
-
-            glow.Delay(duration).FadeOut(400);
-
-            switch (state)
+            if (Hidden)
+                // TODO: figure out a smarter way to include the hit window so we are able to miss but don't keep
+                // it alive unnecessarily long.
+                this.FadeOut(TIME_PREEMPT * FadeOutSpeedMultiplier).Delay(ExpireAfter + 300).Expire();
```
Comment (Aergwyn) [inline osu.Game.Rulesets.Osu/Objects/Drawables/DrawableHitCircle.cs]: I really don't like this `+ 300`. The `.Delay()` is trying to keep it alive so we can click on it before it expires. But that only works up to the `HitObject.StartTime`. Clicking early would work but without the addition you wouldn't be able to miss circles (300ms = hit window). Any ideas to do it better?

Comment (peppy) [reply osu.Game.Rulesets.Osu/Objects/Drawables/DrawableHitCircle.cs]: if anything it should be referencing the hit window value directly. i think it was previously, right?

Comment (Aergwyn) [reply osu.Game.Rulesets.Osu/Objects/Drawables/DrawableHitCircle.cs]: No, not in Lazer. Hit window values were referenced in stable snippets you gave me though. But at different locations.

Comment (Aergwyn) [reply osu.Game.Rulesets.Osu/Objects/Drawables/DrawableHitCircle.cs]: Ah you can access the hit windows. So I could adress this without any constants.

File: osu.Game.Rulesets.Osu/Objects/Drawables/DrawableOsuHitObject.cs [hunk 6]
```diff
@@ -10,9 +10,23 @@ namespace osu.Game.Rulesets.Osu.Objects.Drawables
...
     {
+        /// <summary>
+        /// The number of milliseconds before <see cref="Rulesets.Objects.HitObject.StartTime"/> we should start fading in.
+        /// <para>This should be calculated in the future.</para>
+        /// </summary>
         public const float TIME_PREEMPT = 600;
         public const float TIME_FADEIN = 400;
-        public const float TIME_FADEOUT = 500;
+
+        public bool Hidden;
```
Comment (peppy) [inline osu.Game.Rulesets.Osu/Objects/Drawables/DrawableOsuHitObject.cs]: no

Comment (Aergwyn) [reply osu.Game.Rulesets.Osu/Objects/Drawables/DrawableOsuHitObject.cs]: Is it the name or the bool itself?
Should I go back to the 4 different ones I had before?

Comment (peppy) [reply osu.Game.Rulesets.Osu/Objects/Drawables/DrawableOsuHitObject.cs]: can you list the four you are requiring? generally drawable hitobjects should not be aware of hidden at all (the mod should be doing all the moving)

Comment (Aergwyn) [reply osu.Game.Rulesets.Osu/Objects/Drawables/DrawableOsuHitObject.cs]: I had
`ShowApproachCircle` and `PlayHitAnimation` for DrawableHitCircle,
`FadeOutGradually` for DrawableSlider and
`HideSpinnerDetails` for DrawableSpinner

Comment (peppy) [reply osu.Game.Rulesets.Osu/Objects/Drawables/DrawableOsuHitObject.cs]: Let's go with those for now

Comment (peppy) [reply osu.Game.Rulesets.Osu/Objects/Drawables/DrawableOsuHitObject.cs]: Please make sure to add xmldoc on each explaining what they do, will make it easier to follow for future adventurers

Comment (Aergwyn) [reply osu.Game.Rulesets.Osu/Objects/Drawables/DrawableOsuHitObject.cs]: Going through it only 2 may be sufficient though. 🤔 
ShowApproachCircle and AllowEarlyFadeOut.
The latter would already describe what I need for all 3 types anyways. Also I need something in DrawableOsuHitObject too that allows me to use a different calculation for FadeIn (determined by a bool named `hidden` in stable).
Technically the description would be correct as the HitObjects do fade out earlier. Partially or completely.

Comment (peppy) [reply osu.Game.Rulesets.Osu/Objects/Drawables/DrawableOsuHitObject.cs]: the different calculation should be done by the mod itself, not the drawable implementation

**Comment by Aergwyn** [comment]
Closing this as a better solution is in work and I realized this approach is as wrong as it could be in how fading out is done for the mod.
